File: user_wallet/production/react/src/services/master/SuperAdminService.ts
/**
 * Super Admin Service - React/Web Implementation
 * Identical across ALL platforms
 */

export enum UserRole {
  SUPER_ADMIN = 'super_admin',
  MASTER_ADMIN = 'master_admin',
  WHITE_LABEL_ADMIN = 'white_label_admin',
  USER = 'user',
}

export enum AdminStatus {
  ACTIVE = 'active',
  INACTIVE = 'inactive',
  PENDING = 'pending',
  SUSPENDED = 'suspended',
}

export enum AuthorizationStatus {
  AUTHORIZED = 'authorized',
  PENDING = 'pending',
  REVOKED = 'revoked',
  REJECTED = 'rejected',
}

export interface SuperAdmin {
  id: string;
  email: string;
  passwordHash: string;
  secretKey: string;
  twoFactorEnabled: boolean;
  twoFactorSecret: string;
  phone: string;
  createdAt: number;
  lastLogin: number;
  isActive: boolean;
  permissions: string[];
}

export interface MasterAdmin {
  id: string;
  email: string;
  passwordHash: string;
  authorizedBy: string;
  authorizationStatus: AuthorizationStatus;
  twoFactorEnabled: boolean;
  twoFactorSecret: string;
  phone: string;
  canCreateWhiteLabel: boolean;
  canManageUsers: boolean;
  canManageWallets: boolean;
  canAccessFinance: boolean;
  canModifyFeatures: boolean;
  canManageTokens: boolean;
  canManageNetworks: boolean;
  canViewAnalytics: boolean;
  canManageAdmins: boolean;
  maxWhiteLabels: number;
  whiteLabelCount: number;
  status: AdminStatus;
  createdAt: number;
  lastLogin: number;
  passwordChangedAt: number;
  failedAttempts: number;
  lockedUntil: number;
}

export interface WhiteLabelAdmin {
  id: string;
  email: string;
  passwordHash: string;
  masterAdminId: string;
  brandName: string;
  brandLogo: string;
  brandColor: string;
  customDomain: string;
  authorizationStatus: AuthorizationStatus;
  twoFactorEnabled: boolean;
  twoFactorSecret: string;
  canCustomizeUi: boolean;
  canCustomizeFees: boolean;
  canManageUsers: boolean;
  canManageWallets: boolean;
  canAccessAnalytics: boolean;
  canManageTokens: boolean;
  feePercentage: number;
  status: AdminStatus;
  createdAt: number;
  lastLogin: number;
}

export interface FeatureControl {
  featureName: string;
  enabled: boolean;
  globalEnabled: boolean;
  masterAdminId: string;
  whiteLabelId: string;
  updatedBy: string;
  updatedAt: number;
}

export interface AuditLog {
  id: string;
  adminId: string;
  adminRole: UserRole;
  action: string;
  details: string;
  ipAddress: string;
  userAgent: string;
  timestamp: number;
}

// PROFIT SHARING
export interface ProfitShareConfig {
  id: string;
  whiteLabelId: string;
  superAdminWallet: string;
  masterWalletAddress: string;
  profitPercentage: number;
  minPercentage: number;
  maxPercentage: number;
  isActive: boolean;
  autoTransfer: boolean;
  transferFrequency: string;
  lastTransfer: number;
  totalTransferred: number;
  createdAt: number;
  updatedAt: number;
}

export interface ProfitTransaction {
  id: string;
  whiteLabelId: string;
  superAdminWallet: string;
  amount: number;
  percentage: number;
  grossRevenue: number;
  netRevenue: number;
  token: string;
  txHash: string;
  status: string;
  createdAt: number;
}

class SuperAdminServiceClass {
  private static instance: SuperAdminServiceClass;
  private superAdmins: Map<string, SuperAdmin> = new Map();
  private masterAdmins: Map<string, MasterAdmin> = new Map();
  private whiteLabelAdmins: Map<string, WhiteLabelAdmin> = new Map();
  private featureControls: Map<string, FeatureControl> = new Map();
  private auditLogs: AuditLog[] = [];
  // PROFIT SHARING
  private profitShareConfigs: Map<string, ProfitShareConfig> = new Map();
  private profitTransactions: ProfitTransaction[] = [];

  static getInstance(): SuperAdminServiceClass {
    if (!SuperAdminServiceClass.instance) {
      SuperAdminServiceClass.instance = new SuperAdminServiceClass();
      SuperAdminServiceClass.instance.initialize();
    }
    return SuperAdminServiceClass.instance;
  }

  private initialize(): void {
    this.createDefaultSuperAdmin();
    this.initializeFeatureControls();
  }

  private createDefaultSuperAdmin(): void {
    const superAdmin: SuperAdmin = {
      id: 'super_admin_001',
      email: 'superadmin@tigerwallet.com',
      passwordHash: this.hashPassword(process.env.REACT_APP_SUPER_ADMIN_PASSWORD || ''),
      secretKey: this.generateSecretKey(),
      twoFactorEnabled: false,
      twoFactorSecret: '',
      phone: '',
      createdAt: Date.now(),
      lastLogin: 0,
      isActive: true,
      permissions: ['*'],
    };
    this.superAdmins.set(superAdmin.id, superAdmin);
    this.superAdmins.set(superAdmin.email, superAdmin);
  }

  private initializeFeatureControls(): void {
    const features = [
      'master_wallet_creation', 'multi_blockchain', 'token_management',
      'user_wallet_ownership', 'hd_wallet', 'biometric_auth',
      'pin_code_auth', 'nft_support', 'defi_integration', 'staking',
      'bridge_support', 'mev_protection', 'swap_trading', 'hardware_wallet',
      'admin_controls', 'network_management', 'gas_optimization', 'multi_sig',
      'transaction_history', 'price_alerts', 'privacy_zk', 'coinjoin',
      'account_abstraction', 'session_keys', 'paymaster', 'passkeys',
      'tax_integration', 'analytics', 'cross_chain_intent', 'dapp_browser',
    ];
    features.forEach(feature => {
      this.featureControls.set(feature, {
        featureName: feature,
        enabled: true,
        globalEnabled: true,
        masterAdminId: '',
        whiteLabelId: '',
        updatedBy: '',
        updatedAt: Date.now(),
      });
    });
  }

  // Super Admin Login
  superAdminLogin(email: string, password: string, twoFactorCode: string = ''): SuperAdmin | null {
    const superAdmin = this.superAdmins.get(email);
    if (!superAdmin || !superAdmin.isActive) return null;

    if (this.hashPassword(password) !== superAdmin.passwordHash) {
      this.logAudit(superAdmin.id, UserRole.SUPER_ADMIN, 'LOGIN_FAILED', 'Invalid password');
      return null;
    }

    if (superAdmin.twoFactorEnabled && !this.verifyTwoFactor(superAdmin.twoFactorSecret, twoFactorCode)) {
      return null;
    }

    this.logAudit(superAdmin.id, UserRole.SUPER_ADMIN, 'LOGIN_SUCCESS', 'Super admin logged in');
    return superAdmin;
  }

  // Master Admin Operations
  createMasterAdminRequest(email: string, requestedBy: string): MasterAdmin {
    const masterAdmin: MasterAdmin = {
      id: this.generateId(),
      email,
      passwordHash: this.hashPassword(this.generateTempPassword()),
      authorizedBy: '',
      authorizationStatus: AuthorizationStatus.PENDING,
      twoFactorEnabled: false,
      twoFactorSecret: '',
      phone: '',
      canCreateWhiteLabel: false,
      canManageUsers: false,
      canManageWallets: false,
      canAccessFinance: false,
      canModifyFeatures: false,
      canManageTokens: false,
      canManageNetworks: false,
      canViewAnalytics: false,
      canManageAdmins: false,
      maxWhiteLabels: 0,
      whiteLabelCount: 0,
      status: AdminStatus.PENDING,
      createdAt: Date.now(),
      lastLogin: 0,
      passwordChangedAt: 0,
      failedAttempts: 0,
      lockedUntil: 0,
    };
    this.masterAdmins.set(masterAdmin.id, masterAdmin);
    this.masterAdmins.set(email, masterAdmin);
    this.logAudit('SYSTEM', UserRole.SUPER_ADMIN, 'MASTER_ADMIN_REQUEST', `New request: ${email}`);
    return masterAdmin;
  }

  authorizeMasterAdmin(superAdminId: string, masterAdminId: string, authorized: boolean, notes: string = ''): boolean {
    if (!this.superAdmins.has(superAdminId)) {
      throw new Error('Only super admin can authorize');
    }

    const masterAdmin = this.masterAdmins.get(masterAdminId);
    if (!masterAdmin) return false;

    const updated: MasterAdmin = {
      ...masterAdmin,
      authorizedBy: superAdminId,
      authorizationStatus: authorized ? AuthorizationStatus.AUTHORIZED : AuthorizationStatus.REJECTED,
      status: authorized ? AdminStatus.ACTIVE : AdminStatus.INACTIVE,
    };

    this.masterAdmins.set(masterAdmin.id, updated);
    this.masterAdmins.set(updated.email, updated);

    const action = authorized ? 'AUTHORIZED' : 'REJECTED';
    this.logAudit(superAdminId, UserRole.SUPER_ADMIN, `MASTER_ADMIN_${action}`, `${action} ${masterAdmin.email}`);
    return true;
  }

  masterAdminLogin(email: string, password: string, twoFactorCode: string = ''): MasterAdmin | null {
    const masterAdmin = this.masterAdmins.get(email);
    if (!masterAdmin) return null;

    if (masterAdmin.authorizationStatus !== AuthorizationStatus.AUTHORIZED) return null;
    if (masterAdmin.status !== AdminStatus.ACTIVE) return null;
    if (masterAdmin.lockedUntil > Date.now()) return null;

    if (this.hashPassword(password) !== masterAdmin.passwordHash) {
      this.logAudit(masterAdmin.id, UserRole.MASTER_ADMIN, 'LOGIN_FAILED', 'Invalid password');
      return null;
    }

    if (masterAdmin.twoFactorEnabled && !this.verifyTwoFactor(masterAdmin.twoFactorSecret, twoFactorCode)) {
      return null;
    }

    this.logAudit(masterAdmin.id, UserRole.MASTER_ADMIN, 'LOGIN_SUCCESS', 'Master admin logged in');
    return masterAdmin;
  }

  changeMasterAdminPassword(adminId: string, oldPassword: string, newPassword: string): boolean {
    const masterAdmin = this.masterAdmins.get(adminId) || Array.from(this.masterAdmins.values()).find(m => m.email === adminId);
    if (!masterAdmin) return false;

    if (this.hashPassword(oldPassword) !== masterAdmin.passwordHash) return false;
    if (newPassword.length < 8) return false;

    masterAdmin.passwordHash = this.hashPassword(newPassword);
    masterAdmin.passwordChangedAt = Date.now();
    this.masterAdmins.set(masterAdmin.id, masterAdmin);

    this.logAudit(adminId, UserRole.MASTER_ADMIN, 'PASSWORD_CHANGED', 'Password changed');
    return true;
  }

  enableMasterAdmin2FA(adminId: string, secret: string): boolean {
    const masterAdmin = this.masterAdmins.get(adminId);
    if (!masterAdmin) return false;

    masterAdmin.twoFactorEnabled = true;
    masterAdmin.twoFactorSecret = secret;
    this.masterAdmins.set(masterAdmin.id, masterAdmin);

    this.logAudit(adminId, UserRole.MASTER_ADMIN, '2FA_ENABLED', '2FA enabled');
    return true;
  }

  // White Label Admin
  createWhiteLabelAdmin(masterAdminId: string, email: string, brandName: string): WhiteLabelAdmin | null {
    const masterAdmin = this.masterAdmins.get(masterAdminId);
    if (!masterAdmin || !masterAdmin.canCreateWhiteLabel) return null;
    if (masterAdmin.whiteLabelCount >= masterAdmin.maxWhiteLabels) return null;

    const whiteLabel: WhiteLabelAdmin = {
      id: this.generateId(),
      email,
      passwordHash: this.hashPassword(this.generateTempPassword()),
      masterAdminId,
      brandName,
      brandLogo: '',
      brandColor: '#000000',
      customDomain: '',
      authorizationStatus: AuthorizationStatus.AUTHORIZED,
      twoFactorEnabled: false,
      twoFactorSecret: '',
      canCustomizeUi: true,
      canCustomizeFees: true,
      canManageUsers: true,
      canManageWallets: true,
      canAccessAnalytics: true,
      canManageTokens: true,
      feePercentage: 0,
      status: AdminStatus.ACTIVE,
      createdAt: Date.now(),
      lastLogin: 0,
    };

    this.whiteLabelAdmins.set(whiteLabel.id, whiteLabel);
    this.whiteLabelAdmins.set(email, whiteLabel);

    this.logAudit(masterAdminId, UserRole.MASTER_ADMIN, 'WHITE_LABEL_CREATED', `Created: ${email} - ${brandName}`);
    return whiteLabel;
  }

  // Feature Control
  setGlobalFeature(superAdminId: string, featureName: string, enabled: boolean): boolean {
    if (!this.superAdmins.has(superAdminId)) {
      throw new Error('Only super admin can modify features');
    }

    const feature = this.featureControls.get(featureName);
    if (!feature) return false;

    feature.enabled = enabled;
    feature.globalEnabled = enabled;
    feature.updatedBy = superAdminId;
    feature.updatedAt = Date.now();

    this.logAudit(superAdminId, UserRole.SUPER_ADMIN, 'FEATURE_TOGGLE', `Set ${featureName} = ${enabled}`);
    return true;
  }

  getAllFeatures(): FeatureControl[] {
    return Array.from(this.featureControls.values());
  }

  isFeatureEnabled(featureName: string, adminId: string, role: UserRole): boolean {
    const feature = this.featureControls.get(featureName);
    if (!feature || !feature.globalEnabled) return false;

    switch (role) {
      case UserRole.SUPER_ADMIN:
        return true;
      case UserRole.MASTER_ADMIN:
        if (feature.masterAdminId && feature.masterAdminId !== adminId) return false;
        return feature.enabled;
      case UserRole.WHITE_LABEL_ADMIN:
        if (feature.whiteLabelId && feature.whiteLabelId !== adminId) return false;
        return feature.enabled;
      default:
        return false;
    }
  }

  // Audit
  private logAudit(adminId: string, role: UserRole, action: string, details: string): void {
    const log: AuditLog = {
      id: this.generateId(),
      adminId,
      adminRole: role,
      action,
      details,
      ipAddress: '',
      userAgent: '',
      timestamp: Date.now(),
    };
    this.auditLogs.push(log);
    console.log(`[AUDIT] ${role} | ${action} | ${details}`);
  }

  getAuditLogs(adminId: string = '', limit: number = 100): AuditLog[] {
    if (!adminId) return this.auditLogs.slice(-limit);
    return this.auditLogs.filter(l => l.adminId === adminId).slice(-limit);
  }

  // Helpers — use the Web Crypto API (CSPRNG), never Math.random.
  private generateId(): string { return `id_${Date.now()}_${this.randomHex(4)}`; }
  private generateSecretKey(): string { return this.randomHex(32); }
  private generateTempPassword(): string {
    const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    const bytes = new Uint32Array(16);
    crypto.getRandomValues(bytes);
    return Array.from(bytes, (b) => chars[b % chars.length]).join('');
  }
  private randomHex(byteLength: number): string {
    const bytes = new Uint8Array(byteLength);
    crypto.getRandomValues(bytes);
    return Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
  }
  private hashPassword(password: string): string {
    let hash = 0;
    for (let i = 0; i < password.length; i++) {
      hash = ((hash << 5) - hash + password.charCodeAt(i)) | 0;
    }
    return Math.abs(hash).toString(16);
  }
  private verifyTwoFactor(secret: string, code: string): boolean {
    return code.length === 6 && /^\d+$/.test(code);
  }
}

export const SuperAdminService = SuperAdminServiceClass.getInstance();
export default SuperAdminService;

File: mobile_apps/ios_app/TigerWallet/Master/SuperAdminService.swift
//
//  SuperAdminService.swift
//  TigerWallet
//
//  COMPLETE SUPER ADMIN SYSTEM
//  - Only Super Admin can authorize Master Wallet Admin accounts
//  - Master Wallet Admin login requires Super Admin authorization
//  - Master Wallet Admin can change password and set 2FA after login
//  - Super Admin has FULL control over all features
//  - White Label Admin has full control in their custom branding
//

import Foundation
import CommonCrypto

// MARK: - Enums

enum UserRole: String { case SUPER_ADMIN, MASTER_ADMIN, WHITE_LABEL_ADMIN, USER }
enum AdminStatus: String { case ACTIVE, INACTIVE, PENDING, SUSPENDED }
enum AuthorizationStatus: String { case AUTHORIZED, PENDING, REVOKED, REJECTED }

// MARK: - Data Models

struct SuperAdmin {
    let id: String
    let email: String
    let passwordHash: String
    let secretKey: String
    var twoFactorEnabled: Bool
    var twoFactorSecret: String
    let phone: String
    let createdAt: TimeInterval
    var lastLogin: TimeInterval
    let isActive: Bool
    let permissions: [String]
}

struct MasterAdmin {
    let id: String
    let email: String
    var passwordHash: String
    var authorizedBy: String
    var authorizationStatus: AuthorizationStatus
    var twoFactorEnabled: Bool
    var twoFactorSecret: String
    let phone: String
    var canCreateWhiteLabel: Bool
    var canManageUsers: Bool
    var canManageWallets: Bool
    var canAccessFinance: Bool
    var canModifyFeatures: Bool
    var canManageTokens: Bool
    var canManageNetworks: Bool
    var canViewAnalytics: Bool
    var canManageAdmins: Bool
    var maxWhiteLabels: Int
    var whiteLabelCount: Int
    var status: AdminStatus
    let createdAt: TimeInterval
    var lastLogin: TimeInterval
    var passwordChangedAt: TimeInterval
    var failedAttempts: Int
    var lockedUntil: TimeInterval
}

struct WhiteLabelAdmin {
    let id: String
    let email: String
    var passwordHash: String
    let masterAdminId: String
    var brandName: String
    var brandLogo: String
    var brandColor: String
    var customDomain: String
    var authorizationStatus: AuthorizationStatus
    var twoFactorEnabled: Bool
    var twoFactorSecret: String
    var canCustomizeUi: Bool
    var canCustomizeFees: Bool
    var canManageUsers: Bool
    var canManageWallets: Bool
    var canAccessAnalytics: Bool
    var canManageTokens: Bool
    var feePercentage: Double
    var status: AdminStatus
    let createdAt: TimeInterval
    var lastLogin: TimeInterval
}

struct FeatureControl {
    var featureName: String
    var enabled: Bool
    var globalEnabled: Bool
    var masterAdminId: String
    var whiteLabelId: String
    var updatedBy: String
    var updatedAt: TimeInterval
}

struct AuditLog {
    let id: String
    let adminId: String
    let adminRole: UserRole
    let action: String
    let details: String
    let ipAddress: String
    let userAgent: String
    let timestamp: TimeInterval
}

// MARK: - Super Admin Service

class SuperAdminService {
    static let shared = SuperAdminService()
    
    private var superAdmins: [String: Any] = [:]
    private var masterAdmins: [String: Any] = [:]
    private var whiteLabelAdmins: [String: Any] = [:]
    private var featureControls: [String: FeatureControl] = [:]
    private var auditLogs: [AuditLog] = []
    
    private init() {
        createDefaultSuperAdmin()
        initializeFeatureControls()
    }
    
    // MARK: - Default Super Admin
    
    private func createDefaultSuperAdmin() {
        let superAdmin = SuperAdmin(
            id: "super_admin_001",
            email: "superadmin@tigerwallet.com",
            passwordHash: sha256("SuperAdmin@2024!"),
            secretKey: generateSecretKey(),
            twoFactorEnabled: false,
            twoFactorSecret: "",
            phone: "",
            createdAt: Date().timeIntervalSince1970,
            lastLogin: 0,
            isActive: true,
            permissions: ["*"]
        )
        superAdmins[superAdmin.id] = superAdmin
        superAdmins[superAdmin.email] = superAdmin
    }
    
    private func initializeFeatureControls() {
        let features = ["master_wallet_creation", "multi_blockchain", "token_management", "user_wallet_ownership", "hd_wallet", "biometric_auth", "pin_code_auth", "nft_support", "defi_integration", "staking", "bridge_support", "mev_protection", "swap_trading", "hardware_wallet", "admin_controls", "network_management", "gas_optimization", "multi_sig", "transaction_history", "price_alerts", "privacy_zk", "coinjoin", "account_abstraction", "session_keys", "paymaster", "passkeys", "tax_integration", "analytics", "cross_chain_intent", "dapp_browser"]
        
        for feature in features {
            featureControls[feature] = FeatureControl(
                featureName: feature,
                enabled: true,
                globalEnabled: true,
                masterAdminId: "",
                whiteLabelId: "",
                updatedBy: "",
                updatedAt: Date().timeIntervalSince1970
            )
        }
    }
    
    // MARK: - Super Admin Login
    
    func superAdminLogin(email: String, password: String, twoFactorCode: String = "") -> SuperAdmin? {
        guard let superAdmin = superAdmins[email] as? SuperAdmin else { return nil }
        guard superAdmin.isActive else { return nil }
        
        if sha256(password) != superAdmin.passwordHash {
            logAudit(superAdmin.id, role: .SUPER_ADMIN, action: "LOGIN_FAILED", details: "Invalid password")
            return nil
        }
        
        if superAdmin.twoFactorEnabled && !verifyTwoFactor(superAdmin.twoFactorSecret, code: twoFactorCode) {
            return nil
        }
        
        logAudit(superAdmin.id, role: .SUPER_ADMIN, action: "LOGIN_SUCCESS", details: "Super admin logged in")
        return superAdmin
    }
    
    // MARK: - Master Admin Operations
    
    func createMasterAdminRequest(email: String, requestedBy: String) -> MasterAdmin {
        let masterAdmin = MasterAdmin(
            id: generateId(),
            email: email,
            passwordHash: sha256(generateTempPassword()),
            authorizedBy: "",
            authorizationStatus: .PENDING,
            twoFactorEnabled: false,
            twoFactorSecret: "",
            phone: "",
            canCreateWhiteLabel: false,
            canManageUsers: false,
            canManageWallets: false,
            canAccessFinance: false,
            canModifyFeatures: false,
            canManageTokens: false,
            canManageNetworks: false,
            canViewAnalytics: false,
            canManageAdmins: false,
            maxWhiteLabels: 0,
            whiteLabelCount: 0,
            status: .PENDING,
            createdAt: Date().timeIntervalSince1970,
            lastLogin: 0,
            passwordChangedAt: 0,
            failedAttempts: 0,
            lockedUntil: 0
        )
        
        masterAdmins[masterAdmin.id] = masterAdmin
        masterAdmins[email] = masterAdmin
        logAudit("SYSTEM", role: .SUPER_ADMIN, action: "MASTER_ADMIN_REQUEST", details: "New request: \(email)")
        
        return masterAdmin
    }
    
    func authorizeMasterAdmin(superAdminId: String, masterAdminId: String, authorized: Bool, notes: String = "") throws {
        guard superAdmins[superAdminId] != nil else {
            throw NSError(domain: "SuperAdmin", code: 1, userInfo: [NSLocalizedDescriptionKey: "Only super admin can authorize"])
        }
        
        guard var masterAdmin = masterAdmins[masterAdminId] as? MasterAdmin else { return }
        
        masterAdmin.authorizationStatus = authorized ? .AUTHORIZED : .REJECTED
        masterAdmin.status = authorized ? .ACTIVE : .INACTIVE
        masterAdmin.authorizedBy = superAdminId
        
        masterAdmins[masterAdminId] = masterAdmin
        masterAdmins[masterAdmin.email] = masterAdmin
        
        let action = authorized ? "AUTHORIZED" : "REJECTED"
        logAudit(superAdminId, role: .SUPER_ADMIN, action: "MASTER_ADMIN_\(action)", details: "\(action) \(masterAdmin.email)")
    }
    
    func masterAdminLogin(email: String, password: String, twoFactorCode: String = "") -> MasterAdmin? {
        guard var masterAdmin = masterAdmins[email] as? MasterAdmin else { return nil }
        
        guard masterAdmin.authorizationStatus == .AUTHORIZED else { return nil }
        guard masterAdmin.status == .ACTIVE else { return nil }
        
        if masterAdmin.lockedUntil > Date().timeIntervalSince1970 { return nil }
        
        if sha256(password) != masterAdmin.passwordHash {
            masterAdmin.failedAttempts += 1
            if masterAdmin.failedAttempts >= 5 {
                masterAdmin.lockedUntil = Date().addingTimeInterval(900).timeIntervalSince1970
                masterAdmin.status = .SUSPENDED
            }
            masterAdmins[masterAdmin.id] = masterAdmin
            logAudit(masterAdmin.id, role: .MASTER_ADMIN, action: "LOGIN_FAILED", details: "Invalid password")
            return nil
        }
        
        if masterAdmin.twoFactorEnabled && !verifyTwoFactor(masterAdmin.twoFactorSecret, code: twoFactorCode) {
            return nil
        }
        
        masterAdmin.lastLogin = Date().timeIntervalSince1970
        masterAdmin.failedAttempts = 0
        masterAdmins[masterAdmin.id] = masterAdmin
        masterAdmins[masterAdmin.email] = masterAdmin
        
        logAudit(masterAdmin.id, role: .MASTER_ADMIN, action: "LOGIN_SUCCESS", details: "Master admin logged in")
        
        return masterAdmin
    }
    
    func changeMasterAdminPassword(adminId: String, oldPassword: String, newPassword: String) -> Bool {
        guard var masterAdmin = (masterAdmins[adminId] as? MasterAdmin) ?? (masterAdmins.values.first { ($0 as? MasterAdmin)?.email == adminId } as? MasterAdmin) else {
            return false
        }
        
        guard sha256(oldPassword) == masterAdmin.passwordHash else { return false }
        guard newPassword.count >= 8 else { return false }
        
        masterAdmin.passwordHash = sha256(newPassword)
        masterAdmin.passwordChangedAt = Date().timeIntervalSince1970
        
        masterAdmins[masterAdmin.id] = masterAdmin
        masterAdmins[masterAdmin.email] = masterAdmin
        
        logAudit(adminId, role: .MASTER_ADMIN, action: "PASSWORD_CHANGED", details: "Password changed")
        return true
    }
    
    func enableMasterAdmin2FA(adminId: String, secret: String) -> Bool {
        guard var masterAdmin = masterAdmins[adminId] as? MasterAdmin else { return false }
        
        masterAdmin.twoFactorEnabled = true
        masterAdmin.twoFactorSecret = secret
        
        masterAdmins[masterAdmin.id] = masterAdmin
        masterAdmins[masterAdmin.email] = masterAdmin
        
        logAudit(adminId, role: .MASTER_ADMIN, action: "2FA_ENABLED", details: "2FA enabled")
        return true
    }
    
    // MARK: - White Label Admin
    
    func createWhiteLabelAdmin(masterAdminId: String, email: String, brandName: String) -> WhiteLabelAdmin? {
        guard let masterAdmin = masterAdmins[masterAdminId] as? MasterAdmin else { return nil }
        guard masterAdmin.canCreateWhiteLabel else { return nil }
        guard masterAdmin.whiteLabelCount < masterAdmin.maxWhiteLabels else { return nil }
        
        let whiteLabel = WhiteLabelAdmin(
            id: generateId(),
            email: email,
            passwordHash: sha256(generateTempPassword()),
            masterAdminId: masterAdminId,
            brandName: brandName,
            brandLogo: "",
            brandColor: "#000000",
            customDomain: "",
            authorizationStatus: .AUTHORIZED,
            twoFactorEnabled: false,
            twoFactorSecret: "",
            canCustomizeUi: true,
            canCustomizeFees: true,
            canManageUsers: true,
            canManageWallets: true,
            canAccessAnalytics: true,
            canManageTokens: true,
            feePercentage: 0.0,
            status: .ACTIVE,
            createdAt: Date().timeIntervalSince1970,
            lastLogin: 0
        )
        
        whiteLabelAdmins[whiteLabel.id] = whiteLabel
        whiteLabelAdmins[email] = whiteLabel
        
        logAudit(masterAdminId, role: .MASTER_ADMIN, action: "WHITE_LABEL_CREATED", details: "Created: \(email) - \(brandName)")
        
        return whiteLabel
    }
    
    // MARK: - Feature Control
    
    func setGlobalFeature(superAdminId: String, featureName: String, enabled: Bool) throws {
        guard superAdmins[superAdminId] != nil else {
            throw NSError(domain: "SuperAdmin", code: 1, userInfo: [NSLocalizedDescriptionKey: "Only super admin can modify features"])
        }
        
        guard var feature = featureControls[featureName] else { return }
        
        feature.enabled = enabled
        feature.globalEnabled = enabled
        feature.updatedBy = superAdminId
        feature.updatedAt = Date().timeIntervalSince1970
        
        featureControls[featureName] = feature
        logAudit(superAdminId, role: .SUPER_ADMIN, action: "FEATURE_TOGGLE", details: "Set \(featureName) = \(enabled)")
    }
    
    func getAllFeatures() -> [FeatureControl] {
        return Array(featureControls.values)
    }
    
    func isFeatureEnabled(featureName: String, adminId: String, role: UserRole) -> Bool {
        guard let feature = featureControls[featureName] else { return false }
        guard feature.globalEnabled else { return false }
        
        switch role {
        case .SUPER_ADMIN: return true
        case .MASTER_ADMIN:
            if !feature.masterAdminId.isEmpty && feature.masterAdminId != adminId { return false }
            return feature.enabled
        case .WHITE_LABEL_ADMIN:
            if !feature.whiteLabelId.isEmpty && feature.whiteLabelId != adminId { return false }
            return feature.enabled
        default: return false
        }
    }
    
    // MARK: - Audit
    
    private func logAudit(adminId: String, role: UserRole, action: String, details: String, ipAddress: String = "", userAgent: String = "") {
        let log = AuditLog(
            id: generateId(),
            adminId: adminId,
            adminRole: role,
            action: action,
            details: details,
            ipAddress: ipAddress,
            userAgent: userAgent,
            timestamp: Date().timeIntervalSince1970
        )
        auditLogs.append(log)
        print("[AUDIT] \(role.rawValue) | \(action) | \(details)")
    }
    
    func getAuditLogs(adminId: String = "", limit: Int = 100) -> [AuditLog] {
        if adminId.isEmpty { return Array(auditLogs.suffix(limit)) }
        return auditLogs.filter { $0.adminId == adminId }.suffix(limit).map { $0 }
    }
    
    // MARK: - Helpers
    
    private func generateId() -> String { return "id_\(Int(Date().timeIntervalSince1970))_\(Int.random(in: 0...999999))" }
    private func generateSecretKey() -> String { return (0..<32).map { String(format: "%02x", Int.random(in: 0...255)) }.joined() }
    private func generateTempPassword() -> String { return (0..<16).map { "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"[Int.random(in: 0...61)] }.joined() }
    
    private func sha256(_ input: String) -> String {
        let data = Data(input.utf8)
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes { _ = CC_SHA256($0.baseAddress, CC_LONG(data.count), &hash) }
        return hash.map { String(format: "%02x", $0) }.joined()
    }
    
    /// Real TOTP (RFC 6238) verification using HMAC-SHA1 over the shared
    /// secret and the current 30s time step. Rejects any code that does not
    /// match the HOTP value (±1 step window for clock skew). Never accepts an
    /// arbitrary 6-digit code.
    private func verifyTwoFactor(_ secret: String, code: String) -> Bool {
        guard let key = base32Decode(secret) else { return false }
        let steps = Int(Date().timeIntervalSince1970 / 30)
        // ±1 time-step window to tolerate minor clock skew.
        for skew in -1...1 {
            let counter = UInt64(steps + skew).bigEndian
            let counterData = withUnsafeBytes(of: counter) { Data($0) }
            var hmac = [UInt8](repeating: 0, count: Int(CC_SHA1_DIGEST_LENGTH))
            key.withUnsafeBytes { keyPtr in
                counterData.withUnsafeBytes { ctrPtr in
                    CCHmac(
                        CCHmacAlgorithm(kCCHmacAlgSHA1),
                        keyPtr.baseAddress, key.count,
                        ctrPtr.baseAddress, counterData.count,
                        &hmac
                    )
                }
            }
            let o = Int(hmac[hmac.count - 1] & 0x0f)
            let binary: UInt32 =
                (UInt32(hmac[o] & 0x7f) << 24) |
                (UInt32(hmac[o + 1]) << 16) |
                (UInt32(hmac[o + 2]) << 8) |
                UInt32(hmac[o + 3])
            let otp = String(format: "%06d", binary % 1_000_000)
            if otp == code { return true }
        }
        return false
    }

    /// RFC 4648 Base32 decode (uppercase, no padding required).
    private func base32Decode(_ s: String) -> Data? {
        let alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
        var bits = 0
        var value = 0
        var out = Data()
        for ch in s.uppercased() where ch != "=" {
            guard let idx = alphabet.firstIndex(of: ch) else { return nil }
            value = (value << 5) | alphabet.distance(from: alphabet.startIndex, to: idx)
            bits += 5
            if bits >= 8 {
                bits -= 8
                out.append(UInt8((value >> bits) & 0xff))
            }
        }
        return out.isEmpty ? nil : out
    }
}
